### ISSUE TITLE
@nosideeffect extern methods should trigger a warning when their result is not used

### ISSUE THREAD

**Issue by MatrixFrog** [google/closure-compiler#1108]
Idea from @pauldraper 

If a function is `@nosideeffects` and also has no `@return` statement then it doesn't actually *do* anything. This is almost certainly a mistake in the JSDoc annotations.

**Comment by concavelenz** [comment]
`/** @nosideeffects */ function noop() {}; `    :-)

**Comment by pauldraper** [comment]
After thinking about this again, I do see now that there are `angular.noop` and `$.noop`, for passing in as callbacks, which makes sense. (`noop()` in code is utterly worthless, but passing `noop` as an argument is reasonable.)

The point was to prevent a bug like https://github.com/google/closure-compiler/pull/1107.
Removing chunks of code can be confounding, even more so if it is a standard HTML5 function, and bundled with the compiler.

One solution may be to add `@return {undefined}` in the case of an intentional noop:

```javascript
/**
 * @nosideeffects
 * @return {undefined}
 */
angular.noop = function() {};
```

**Comment by ChadKillingsworth** [comment]
The original test case in #1107 should have been flagged as suspicious code by the CheckSideEffects pass. The fact that it wasn't is a bug.

**Comment by ChadKillingsworth** [comment]
The CheckSideEffects pass runs before type checking and therefore cannot use type information currently. We'll either need to move it after the type checking/inference passes or run it a second time later to use type information.

**Comment by ChadKillingsworth** [comment]
Here's the current test that already describes the functionality: https://github.com/google/closure-compiler/blob/master/test/com/google/javascript/jscomp/CheckSideEffectsTest.java#L217

**Comment by igrep** [comment]
I often use `@nosideeffects` in a constructor, which usually has no `return`.
But when the constructor doesn't do anything except initializing `this`, I like to add `@nosideeffects`.
Do you think it's a wrong way?

**Comment by ChadKillingsworth** [comment]
@igrep In externs right? The compiler only users `@nosideeffects` annotations from extern files.

If so, it's not wrong. I'm not sure it would have any effect. `@return` annotations on a constructor specifically designate that the constructor is allowed to be called statically (without the `new` keyword) and what value it returns under those conditions.

**Comment by igrep** [comment]
> The compiler only users @nosideeffects annotations from extern files.

Oh, I didn't know that. :astonished: Thanks.

**Comment by ChadKillingsworth** [comment]
Based on the discussion, I'm not sure we are going to change the existing behavior.
